src/controller_examples/zookeeper_controller/proof/liveness.rs

```diff
@@ -1217,17 +922,129 @@ proof fn lemma_true_leads_to_always_current_state_matches_zk_under_eventual_inva
         spec,
         true_pred(),
         lift_state(|s: ClusterState| !s.reconcile_state_contains(zk.object_ref())),
-        lift_state(at_init_step_with_no_pending_req(zk)),
-        lift_state(at_after_create_headless_service_step_with_zk_and_pending_req_in_flight(zk)),
-        lift_state(at_after_create_client_service_step_with_zk_and_pending_req_in_flight(zk)),
-        lift_state(at_after_create_admin_server_service_step_with_zk_and_pending_req_in_flight(zk)),
-        lift_state(at_after_create_config_map_step_with_zk_and_pending_req_in_flight(zk)),
-        lift_state(at_after_get_stateful_set_step_with_zk_and_pending_req_in_flight(zk)),
+        lift_state(no_pending_req_at_zookeeper_step_with_zk(zk, ZookeeperReconcileStep::Init)),
+        lift_state(pending_req_in_flight_at_zookeeper_step_with_zk(ZookeeperReconcileStep::AfterCreateHeadlessService, zk, arbitrary())),
+        lift_state(pending_req_in_flight_at_zookeeper_step_with_zk(ZookeeperReconcileStep::AfterCreateClientService, zk, arbitrary())),
+        lift_state(pending_req_in_flight_at_zookeeper_step_with_zk(ZookeeperReconcileStep::AfterCreateAdminServerService, zk, arbitrary())),
+        lift_state(pending_req_in_flight_at_zookeeper_step_with_zk(ZookeeperReconcileStep::AfterCreateConfigMap, zk, arbitrary())),
+        lift_state(pending_req_in_flight_at_zookeeper_step_with_zk(ZookeeperReconcileStep::AfterGetStatefulSet, zk, arbitrary())),
         lift_state(current_state_matches(zk)),
         always(lift_state(current_state_matches(zk)))
     );
 }
 
+// This lemma ensures that zookeeper controller at some step with pending request in flight will finally enter its next step.
+// For the initial state and final state, they both require the reconcile_state to have a pending request which is the correct
+// request for that step (parameter 'step' for the initial step, parameter 'next_step' for the final state).
+// Note that in this lemma we add some constraints to the initial step:
+//    1. The next step of it when conducting reconcile_core is deterministic.
```

Comment:
Instead of deterministic, I think what you mean is that there is only one possible next step (i.e., step b) for step a. The reconcile_core is still deterministic even with branches.